utf8/reader_test.go

```diff
@@ -0,0 +1,288 @@
+
+func TestUTF8Reader(t *testing.T) {
+	type testcase struct {
+		input  string
+		repeat int
+		err    error
+		errOut []rune
+	}
+
+	testcases := []testcase{
+		{
+			input:  "",
+			repeat: 1,
+		},
+		{
+			input:  string([]byte{0x80}),
+			repeat: 1,
+			err:    utf8.ErrDecode,
+		},
+		{
+			input:  "test",
+			repeat: 1,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1,
+		},
+		{
+			input:  "Εν οίδα ότι TESTE ουδέν οίδα",
+			repeat: 1,
+		},
+		{
+			input:  "test\n",
+			repeat: 1024,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1024,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1025,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 2048,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 2049,
+		},
+		{
+			input:  string([]byte{206, 149, 206, 206, 206}),
+			repeat: 1,
+			err:    utf8.ErrDecode,
+			errOut: []rune{'Ε'},
+		},
+	}
+
+	// test NewReaderReader
+	for _, tc := range testcases {
+		reader := utf8.NewReaderReader(iotest.NewRepeatReader(
+			bytes.NewBuffer([]byte(tc.input)),
+			tc.repeat))
+
+		got, readErr := runes.ReadAll(reader)
+		assert.IsError(t, tc.err, readErr, "Read() error")
+
+		var expected []rune
+
+		if readErr == nil {
+			repeater := iotest.NewRepeatReader(bytes.NewBuffer([]byte(tc.input)),
+				tc.repeat)
+
+			expectedBytes, err := io.ReadAll(repeater)
+			assert.NoError(t, err, "repeating expected")
+
+			expected = []rune(string(expectedBytes))
+		} else {
+			expected = tc.errOut
+		}
+
+		assert.EqualInts(t, len(expected), len(got), "rune slice len mismatch: %s",
+			string(got))
+
+		for i, r := range expected {
+			if r != got[i] {
+				t.Errorf("want[%d = %c] but got[%d = %c]", r, r, got[i], got[i])
+			}
+		}
+	}
+
+	// test NewReader
+	for _, tc := range testcases {
+		inputReader := iotest.NewRepeatReader(
+			bytes.NewBuffer([]byte(tc.input)),
+			tc.repeat)
+
+		data, err := io.ReadAll(inputReader)
+		assert.NoError(t, err, "reading repeated string")
+
+		reader := utf8.NewReader(bytes.NewBuffer(data))
+
+		got, readErr := runes.ReadAll(reader)
+		assert.IsError(t, tc.err, readErr, "read() error")
+
+		var expected []rune
+
+		if readErr == nil {
+			repeater := iotest.NewRepeatReader(bytes.NewBuffer([]byte(tc.input)),
+				tc.repeat)
+
+			expectedBytes, err := io.ReadAll(repeater)
+			assert.NoError(t, err, "repeating expected")
+
+			expected = []rune(string(expectedBytes))
+		} else {
+			expected = tc.errOut
+		}
+
+		assert.EqualInts(t, len(expected), len(got), "rune slice len mismatch")
+
+		for i, r := range expected {
+			if r != got[i] {
+				t.Errorf("want[%d = %c] but got[%d = %c]", r, r, got[i], got[i])
+			}
+		}
+	}
+
+	// TODO(i4k): implement stdiotest.TestReader for utf8.Reader
+}
+
+func TestUTF8ReaderMultipleSizes(t *testing.T) {
+	input := iotest.NewRepeatReader(bytes.NewBuffer([]byte(expectedString)), 100)
+	inputBytes, err := io.ReadAll(input)
+	assert.NoError(t, err, "failed to repeat input")
```

Comment:
```suggestion
	assert.NoError(t, err)
```

Thanks to t.Helper() when this fails it will be fairly easy to see that it was a io.ReadAll error on the input (it fails on this line, not on the helper assert func), so the detail here doesn't add much IMHO.

Reply:
it states the error is in the repeater.

Reply:
Well it says failed to read input, which when it fails it will fail on the line of code bellow of where input is being read (io.ReadAll(input))...so still not so sure it helps...but it also doesn't hurt, so all good :100:

Reply:
> Well it says failed to read input

it says: failed to **repeat** input... because Read() calls the Repeater Read() so if it fails, it's the repeater failing ... so this is the message.

Reply:
OMG man ..for some reason I was reading "failed to read input" o.O. Sorry for th enoise.